WIP: UTC Timing

julialindberg opened the pull request:
Hello again,

We're looking for feedback on the following code and issue.  

We noticed that the current implementation ignores UTCTiming element from the manifest. This is a problem for live content on devices with wrong time settings. For example, if the device clock is ahead of actual time by an hour, the player will calculate wrong initial time and fail. (https://gitter.im/canalplus/rx-player?at=5c4744bb0721b912a5bf14f4)

We parse `UTCTiming` in `node_parsers/MPD.ts` using `parseScheme`. The url is then fetched using `needs-ressources`, and `schemeIdUri` is converted from `urn:mpeg:dash:utc:http-iso:2014` (with value being an url) to `urn:mpeg:dash:utc:direct:2014` (with value being the date string).

The timing is then used in `manifest.ts` by calculating the offset between server time and the client time.

Known issues:
- Only one UTCTiming is used. (only first one is used)
- The time is fetched every time the manifest is refreshed, which is probably not required
- `needs-ressources` is maybe not the ideal place to fetch UTC timings
- Unsure about the changes in `get_presentation_live_gap.ts`

Thank you!

peaBerberian commented:
Hi and thanks!

I'm on board with using only the first UTCTiming element and ``needs-ressources`` could seem adapted: its usage mainly allow the loader-management logic (such as retries with backoff if the request fails and better errors) to be used through the parser (which completely stop in case of an error). Using it for fetching a time is fine with me.

``IParsedMPD`` is indeed not used anymore, it was before when the parser parsed more attributes than what was needed. We can remove it.

This code is nice as it respects pretty well where code should be updated but before we merge, I prefer to perform some adjustments (that might be a lot of modifications you don't need, so I will try to work on it next week):
  - this request should only be done if it is required (it is not if we have a SegmentTimeline in the MPD or if we are in refreshing the manifest)
  - I'm personally not a fan of calculating the offset from ``Date.now`` in ``Manifest``, as we could create a ``new Manifest()`` at any point in time (for us, it's just an object to facilitate manifest management). Also,  the scheme (``schemeIdUri`` and ``value``) is too much of a DASH concept to be there. Our solution might be to provide a 'clockOffset` coming directly from the parser.
  - for ``get_presentation_live_gap.ts`` the code updated makes sense (even if `lastRef` coincide with the fact that we have a last time reference and thus do not need a clock).
The default `10` value though should be unneeded now (or perhaps decreased) at it was only here as a security if the clock were not perfectly synchronized
  - for the moment, only ``http-iso`` is managed. As it's not a big investment, we might also add ``http-head`` and ``direct``

julialindberg commented:
Thank you for feedback! We will look into the first two bullet points.

julialindberg commented:
We have now done some improvements according to your comments. We moved the clockOffset calculation to the parser and fixed the lint issues. Regarding the first bullet point, do you have any suggestions on how to implement this? We're thinking, that parse_mpd doesn't know whether it has been fetched before or not. Maybe we can pass some parsing options or can you think of a better way?

Unfortunately we don't have much more time this week to look into the last two points. Feel free to continue that work.

peaBerberian commented:
> Maybe we can pass some parsing options or can you think of a better way?

Yes, that's the best solution I can think of for now.

> Unfortunately we don't have much more time this week to look into the last two points. Feel free to continue that work.

OK thanks and no problem. I will try to work on it later this week.

Thanks again for this PR.
It looks like the feature is already functional, I just want to perform those minor improvements once and for all to have a solid and generic UTCTiming management on the corresponding release.

julialindberg commented:
Now utcTiming is only parsed the first time the manifest is fetched. We solved that by passing options to the manifest parsing as we talked about. 

The commit history is a little messy though. Would you like us to fix that for you?

peaBerberian commented:
Hi,

Sorry for the delay, we had some other unrelated issues this month.
I think we are almost good but some little things bothered me:
  - the fact that we leak the word utcTimings" to the core, instead of just something generic like "clock synchronization"
  - `core/init` assumed that we should never need to fetch the clock on refresh, instead of doing it only if we don't have a synchronized clock (even if it's the same final result here).
  - the fact that we fetched UTCTimings even if it was not needed (e.g. for SegmentTimeline-based contents)

I tried to work on it a little this week and I made another branch:
https://github.com/canalplus/rx-player/compare/feat/utc-timing

I still have some issues with it (mostly style-related) but it should be functional. Is it good for your needs?

julialindberg commented:
Hi,

Thank you for your work! Seems to be good improvements.
We will test the new branch to make sure that it fits our needs. 

Thanks!
